### PR TITLE
Attention mask is important in the case of batching...

### LINKED ISSUE #16221: token classification pipeline does not use attention mask, affects predictions
## Environment info

- `transformers` version: 4.17.0
- Platform: Mac
- Python version: 3.7.10
- PyTorch version (GPU?): 1.8.1 no GPU
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help
@Narsil 

## Information

Model I am using: Bert for token classification

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [X] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [X] my own task or dataset: (give details below)

NER pipeline

## To reproduce

```python
from transformers import AutoModelForTokenClassification, AutoTokenizer, \
    TokenClassificationPipeline

model = AutoModelForTokenClassification.from_pretrained("alvaroalon2/biobert_diseases_ner")
tokenizer = AutoTokenizer.from_pretrained("alvaroalon2/biobert_diseases_ner")

pipeline = TokenClassificationPipeline(model=model,
                                       tokenizer=tokenizer,
                                       aggregation_strategy="first",
                                       ignore_labels=["0", "O"],
                                       )
sentences = ["hello, I have influenza",
             "Hepatitis is a known problem for many nations, particularly in "
             "combination with rabies"]

print(pipeline(sentences, batch_size=1))
print(pipeline(sentences, batch_size=2))
```

output:
```python
[[{'entity_group': 'DISEASE', 'score': 0.9999732, 'word': 'influenza', 'start': 14, 'end': 23}], [{'entity_group': 'DISEASE', 'score': 0.99998534, 'word': 'Hepatitis', 'start': 0, 'end': 9}, {'entity_group': 'DISEASE', 'score': 0.99989414, 'word': 'rabies', 'start': 80, 'end': 86}]]
[[], [{'entity_group': 'DISEASE', 'score': 0.99998534, 'word': 'Hepatitis', 'start': 0, 'end': 9}, {'entity_group': 'DISEASE', 'score': 0.99989414, 'word': 'rabies', 'start': 80, 'end': 86}]]
```

## Expected behavior

The outputs should be the same regardless of batch size. The attention mask is not passed to the model, leading to the discrepancy. If may just be a case of setting `return_attention_mask=True` in the [preprocessing](https://github.com/huggingface/transformers/blob/25b8f9a85b929e2b9acb190b6df7f094d07e4b61/src/transformers/pipelines/token_classification.py#L195) function.

Let me know if you need any more clarification!

Thanks

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes #16221

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/base.py [hunk 1]
```diff
@@ -149,7 +149,7 @@ def inner(items):
                 _padding_value = t_padding_value
             elif key in {"input_values", "pixel_values", "input_features"}:
                 _padding_value = f_padding_value
-            elif key in {"p_mask"}:
+            elif key in {"p_mask", "special_tokens_mask"}:
```
Comment (Narsil) [inline src/transformers/pipelines/base.py:152]: In random models, `special_tokens_mask` would be extended in the batch with `0` instead of `1` so we could still predict `PAD` token in the pipeline.

I think having `pad` being always considered a `special_tokens_mask` is fine.

File: src/transformers/pipelines/token_classification.py [hunk 2]
```diff
@@ -192,7 +192,6 @@ def preprocess(self, sentence, offset_mapping=None):
         truncation = True if self.tokenizer.model_max_length and self.tokenizer.model_max_length > 0 else False
         model_inputs = self.tokenizer(
             sentence,
-            return_attention_mask=False,
```
Comment (Narsil) [inline src/transformers/pipelines/token_classification.py:195]: `return_attention_mask=True` is also incorrect because FNet doesn't expect an attention mask.

Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:195]: Maybe FNet will continue to exhibit the flaw that pad tokens modify the output, I don't know enough about it though.

Comment (sgugger) [reply src/transformers/pipelines/token_classification.py:195]: You will thus get the attention mask since you don't remove it afterward, but I'm guessing that's the whole point?

**Review by sgugger — APPROVED** [review]
LGTM, thanks for fixing!

File: src/transformers/pipelines/token_classification.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:195]: Actually it seems the FNet tokenizer doesn't return attention mask if we don't ask  for it. (Which is fair since the model doesn't seem to accept them).
